addOns/ascanrules/src/main/java/org/zaproxy/zap/extension/ascanrules/parserapi/ParserAPI.java

```diff
@@ -0,0 +1,142 @@
+/*
+ * Zed Attack Proxy (ZAP) and its related class files.
+ *
+ * ZAP is an HTTP/HTTPS proxy for assessing web application security.
+ *
+ * Copyright 2022 The ZAP Development Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.zaproxy.zap.extension.ascanrules.parserapi;
+
+import static org.zaproxy.zap.extension.ascanrules.utils.Constants.CONTEXT;
+import static org.zaproxy.zap.extension.ascanrules.utils.Constants.DOUBLE_QUOTE_CHARACTER;
+import static org.zaproxy.zap.extension.ascanrules.utils.Constants.FORWARD_SLASH_CHARACTER;
+import static org.zaproxy.zap.extension.ascanrules.utils.Constants.SINGLE_QUOTE_CHARACTER;
+
+import java.io.IOException;
+import java.util.List;
+import net.htmlparser.jericho.Element;
+import net.htmlparser.jericho.HTMLElementName;
+import net.htmlparser.jericho.Source;
+import org.antlr.v4.runtime.BaseErrorListener;
+import org.antlr.v4.runtime.CharStream;
+import org.antlr.v4.runtime.CharStreams;
+import org.antlr.v4.runtime.CommonTokenStream;
+import org.antlr.v4.runtime.RecognitionException;
+import org.antlr.v4.runtime.Recognizer;
+import org.antlr.v4.runtime.Token;
+import org.parosproxy.paros.network.HttpMessage;
+import org.zaproxy.zap.extension.ascanrules.parserapi.impl.JavaScriptLexer;
+import org.zaproxy.zap.extension.ascanrules.parserapi.impl.JavaScriptParser;
+
+class MyErrorListener extends BaseErrorListener {
+    @Override
+    public void syntaxError(
+            Recognizer<?, ?> recognizer,
+            Object offendingSymbol,
+            int line,
+            int charPositionInLine,
+            String msg,
+            RecognitionException e) {
+        throw new java.lang.Error("Syntax Error");
+    }
+}
+
+public class ParserAPI {
+
+    private static String scriptCode;
+    private static int targetBlockNumber = 0;
```

Comment:
Should not be static, the scan rules can be executed concurrently which would lead to mixed state.

Reply:
In order to make this not static I had to add a ParserApi member to the scan rule. I hope that's what you intended.

Reply:
Yes but non-static and better instantiate in the `init` method.

Reply:
The `scriptCode` shouldn't be static either (the comment was added to the lines 58 and 59 ;)

Reply:
I didn't review the whole flow, maybe it should be initialised on/before each attack.

Reply:
> Yes but non-static and better instantiate in the init method.

I did remove the static modifier but haven't moved it to init yet. Will do that in a bit.

Reply:
Worth checking the logic first, might have to be done elsewhere.